cmd/formatter/shortcut.go

```diff
@@ -0,0 +1,324 @@
+	if ke.shoudlDisplay() {
+		errMessage := ke.err.Error()
+
+		MoveCursor(height-linesOffset(info)-linesOffset(errMessage)-1, 0)
+		ClearLine()
+
+		fmt.Print(errMessage)
+	}
+}
+
+func (ke *KeyboardError) addError(prefix string, err error) {
+	ke.timeStart = time.Now()
+
+	prefix = ansiColor(CYAN, fmt.Sprintf("%s →", prefix), BOLD)
+	errorString := fmt.Sprintf("%s  %s", prefix, err.Error())
+
+	ke.err = errors.New(errorString)
+}
+
+func (ke *KeyboardError) error() string {
+	return ke.err.Error()
+}
+
+type KeyboardWatch struct {
+	Watcher  watch.Notify
+	Watching bool
+	WatchFn  func(ctx context.Context, project *types.Project, services []string, options api.WatchOptions) error
+	Ctx      context.Context
+	Cancel   context.CancelFunc
+}
+
+func (kw *KeyboardWatch) isWatching() bool {
+	return kw.Watching
+}
+
+func (kw *KeyboardWatch) switchWatching() {
+	kw.Watching = !kw.Watching
+}
+
+func (kw *KeyboardWatch) newContext(ctx context.Context) context.CancelFunc {
+	ctx, cancel := context.WithCancel(ctx)
+	kw.Ctx = ctx
+	kw.Cancel = cancel
+	return cancel
+}
+
+type KEYBOARD_LOG_LEVEL int
+
+const (
+	NONE  KEYBOARD_LOG_LEVEL = 0
+	INFO  KEYBOARD_LOG_LEVEL = 1
+	DEBUG KEYBOARD_LOG_LEVEL = 2
+)
+
+type LogKeyboard struct {
+	kError                KeyboardError
+	Watch                 KeyboardWatch
+	IsDockerDesktopActive bool
+	IsWatchConfigured     bool
+	logLevel              KEYBOARD_LOG_LEVEL
+	signalChannel         chan<- os.Signal
+}
+
+var KeyboardManager *LogKeyboard
+var eg multierror.Group
+
+func NewKeyboardManager(ctx context.Context, isDockerDesktopActive, isWatchConfigured bool,
+	sc chan<- os.Signal,
+	watchFn func(ctx context.Context,
+		project *types.Project,
+		services []string,
+		options api.WatchOptions,
+	) error,
+) {
+	km := LogKeyboard{}
+	km.IsDockerDesktopActive = isDockerDesktopActive
+	km.IsWatchConfigured = isWatchConfigured
+	km.logLevel = INFO
+
+	km.Watch.Watching = false
+	km.Watch.WatchFn = watchFn
+
+	km.signalChannel = sc
+
+	KeyboardManager = &km
+
+	HideCursor()
+}
+
+func (lk *LogKeyboard) PrintKeyboardInfo(printFn func()) {
+	lk.clearNavigationMenu()
+	printFn()
+
+	if lk.logLevel == INFO {
+		lk.createBuffer(0)
+		lk.printNavigationMenu()
+	}
+}
+
+// Creates space to print error and menu string
+func (lk *LogKeyboard) createBuffer(lines int) {
+	allocateSpace(lines)
+
+	if lk.kError.shoudlDisplay() {
+		extraLines := linesOffset(lk.kError.error()) + 1
+		allocateSpace(extraLines)
+		lines += extraLines
+	}
+
+	infoMessage := lk.navigationMenu()
+	extraLines := linesOffset(infoMessage) + 1
+	allocateSpace(extraLines)
+	lines += extraLines
+
+	if lines > 0 {
+		MoveCursorUp(lines)
+	}
+}
+
+func (lk *LogKeyboard) printNavigationMenu() {
+	if lk.logLevel == INFO {
+		height := goterm.Height()
+		menu := lk.navigationMenu()
+
+		MoveCursorX(0)
+		SaveCursor()
+
+		lk.kError.printError(height, menu)
+
+		MoveCursor(height-linesOffset(menu), 0)
+		ClearLine()
+		fmt.Print(menu)
```

Comment:
(for follow-up)

You can make it easier to test this by storing stdout/`io.Writer` on the type and writing to that.

Reply:
you mean, instead of using fmt.Print first pipe all the characters to stdout and then print?

Reply:
add an `io.Writer` field in `KeyboardError` type, and use `fmt.Fprint`. This will allow unit tests to pass a `bytes.Buffer` as writer and check output
For actual use, should be initialized to `dockercli.Out()` to respect docker CLI plugin architecture